lib/cn/cn.c

```diff
@@ -1099,17 +1094,27 @@ cn_open(
     cn->cp = kvdb_kvs_cparams(kvs);
     cn->cn_cndb = cndb;
     cn->ikvdb = ikvdb_kvdb_handle(kvdb_kvs_parent(kvs));
-    cn->cn_dataset = ds;
+    cn->cn_dataset = mp;
     cn->cn_cnid = cnid;
     cn->cn_cflags = kvdb_kvs_flags(kvs);
     cn->cn_kvdb_health = health;
     cn->cn_mpool_props = mpprops;
 
-    staging_absent = mpool_mclass_props_get(ds, MP_MED_STAGING, NULL);
-    if (staging_absent) {
-        if (strcmp(rp->mclass_policy, "capacity_only")) {
-            log_warn("Staging media not configured, using capacity_only media class policy");
-            strlcpy(rp->mclass_policy, "capacity_only", HSE_MPOLICY_NAME_LEN_MAX);
+    if (ikvdb_pmem_only_get(cn->ikvdb)) {
+        log_info("KVDB is pmem-only, changing the mclass policy to \"pmem_only\" for the KVS %s",
+                 kvs_name);
+        strlcpy(rp->mclass_policy, "pmem_only", HSE_MPOLICY_NAME_LEN_MAX);
+    }
```

Comment:
I'd like to hoist this logic out of cn_open.  Move it up the stack and then you don't need the API `ikvdb_pmem_only_get`.  May not even need to store `pmem_only` in the ikvdb struct.

Reply:
Agreed, will hoist this out of cn_open() and remove ikvdb_pmem_only_get().

I might still have to retain the "pmem_only" bool in ikvdb_impl so that KVS open need not re-evaluate it.

Reply:
It is expensive to re-eval, so it is worth caching.  But if it's only used to set the mclass policy and to fail mpool create/open, then it doesn't really need to be cached.  Use it and lose it.

Reply:
KVDB open evaluates it and any KVS open that follows retrieves it from the kvdb handle.

In the context of KVS open, it is *only* used to override the mclass policy to "pmem_only" if the KVDB is a pmem-only KVDB.